File: macos/PackMan/Sources/PackMan/PackageSources/PipxSource.swift
import Foundation

struct PipxSource: PackageSource {
    let runner: any ProcessRunning
    let resolver: any ToolResolving
    let httpClient: any HTTPDataLoading
    private let capabilities = PipxCapabilities()

    init(
        runner: any ProcessRunning = ProcessRunner.shared,
        resolver: any ToolResolving = ToolResolver.shared,
        httpClient: any HTTPDataLoading = URLSessionHTTPClient()
    ) {
        self.runner = runner
        self.resolver = resolver
        self.httpClient = httpClient
    }

    let descriptor = SourceDescriptor(
        id: .pipx,
        name: "pipx",
        toolID: .pipx,
        executableName: "pipx",
        knownPaths: ["/opt/homebrew/bin/pipx", "/usr/local/bin/pipx"],
        installationURL: URL(string: "https://pipx.pypa.io/stable/installation/"))

    func probe() async -> SourceProbe {
        await SourceSupport.probe(
            descriptor: descriptor,
            versionArguments: ["--version"],
            resolver: resolver,
            runner: runner)
    }

    func scan(
        context: ToolContext,
        progress: @escaping @Sendable (SourcePhase) async -> Void
    ) async throws -> SourceScanReport {
        await progress(.scanning)
        if try await supportsNativeOutdated(context: context) {
            return try await scanNative(context: context)
        }
        return try await scanLegacy(context: context)
    }

    func update(
        request: UpdateRequest,
        context: ToolContext,
        onOutput: @escaping @Sendable (ProcessOutputEvent) async -> Void
    ) async throws {
        guard PackageIdValidator.isValid(request.packageID) else {
            throw SourceError.invalidPackageId(request.packageID)
        }
        let result = try await runner.run(
            context.executablePath,
            ["upgrade", request.packageID],
            timeout: 600,
            environment: SourceSupport.environment(pathEntries: context.pathEntries),
            onOutput: onOutput)
        guard result.succeeded else { throw SourceSupport.commandFailure("pipx upgrade", result: result) }
    }

    private func supportsNativeOutdated(context: ToolContext) async throws -> Bool {
        try await capabilities.supportsNativeOutdated(context: context, runner: runner)
    }

    private func scanNative(context: ToolContext) async throws -> SourceScanReport {
        let result = try await runner.run(
            context.executablePath,
            ["list", "--outdated", "--output", "json"],
            timeout: 180,
            environment: SourceSupport.environment(pathEntries: context.pathEntries))
        guard let data = result.stdout.data(using: .utf8), !result.stdout.trimmed.isEmpty else {
            if result.succeeded { throw SourceError.commandFailed("pipx returned empty output instead of JSON.") }
            throw SourceSupport.commandFailure("pipx list --outdated", result: result)
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let envelope: PipxOutdatedEnvelope
        do {
            envelope = try decoder.decode(PipxOutdatedEnvelope.self, from: data)
        } catch {
            throw SourceError.commandFailed("pipx outdated JSON parse failed: \(error.decodingDescription)")
        }

        let updates = envelope.data.packages
            .filter { !$0.injected && !$0.pinned && PackageIdValidator.isValid($0.environment) && PackageIdValidator.isValid($0.package) && PackageIdValidator.isValidVersion($0.latestVersion) }
            .map {
                PackageInfo(
                    id: $0.environment,
                    name: $0.environment == $0.package ? $0.package : "\($0.package) (\($0.environment))",
                    currentVersion: $0.version,
                    availableVersion: $0.latestVersion, registryID: $0.package)
            }
        var issues = envelope.errors.map { error in
            let scope = error.environment ?? error.package
            return SourceIssue(
                kind: .network,
                message: scope.map { "\($0): \(error.message)" } ?? error.message,
                recovery: "Check the package index configuration and retry.")
        }
        if (!result.succeeded || envelope.exitCode != 0 || envelope.status != "success") && issues.isEmpty {
            issues.append(SourceIssue(
                kind: .command,
                message: "pipx reported an unsuccessful outdated check.",
                recovery: result.stderr.trimmed.isEmpty ? "Retry the scan." : result.stderr.trimmed))
        }
        let invalid = envelope.data.packages.filter { !PackageIdValidator.isValid($0.environment) || !PackageIdValidator.isValid($0.package) || !PackageIdValidator.isValidVersion($0.latestVersion) }
        if !invalid.isEmpty { issues.append(SourceIssue(kind: .parsing, message: "pipx returned invalid package records.")) }
        return SourceScanReport(updates: updates, issues: issues,
            skippedCount: envelope.data.skipped.count + envelope.data.packages.filter { $0.pinned || $0.injected }.count)
    }

    private func scanLegacy(context: ToolContext) async throws -> SourceScanReport {
        let result = try await runner.run(
            context.executablePath,
            ["list", "--short"],
            timeout: 60,
            environment: SourceSupport.environment(pathEntries: context.pathEntries))
        guard result.succeeded else { throw SourceSupport.commandFailure("pipx list", result: result) }

        var installed: [(name: String, version: String)] = []
        var issues: [SourceIssue] = [SourceIssue(kind: .configuration, message: "Legacy pipx uses public PyPI estimates; private indexes, constraints, and pins cannot be checked reliably. Upgrade pipx for native checks.")]
        for line in result.stdout.split(separator: "\n") {
            let parts = line.split(separator: " ", omittingEmptySubsequences: true)
            guard parts.count >= 2, PackageIdValidator.isValid(String(parts[0])) else {
                issues.append(SourceIssue(kind: .parsing, message: "Could not parse a pipx package record."))
                continue
            }
            installed.append((String(parts[0]), String(parts[1])))
        }

        let lookups = await withTaskGroup(of: PipxLookup.self, returning: [PipxLookup].self) { group in
            var iterator = installed.makeIterator()
            var results: [PipxLookup] = []

            func enqueue() {
                guard !Task.isCancelled, let package = iterator.next() else { return }
                group.addTask {
                    await lookupPyPI(name: package.name, currentVersion: package.version)
                }
            }

            for _ in 0..<min(8, installed.count) { enqueue() }
            while let lookup = await group.next() {
                results.append(lookup)
                enqueue()
            }
            return results
        }

        var updates: [PackageInfo] = []
        for lookup in lookups {
            switch lookup {
            case .current:
                break
            case .outdated(let info):
                updates.append(info)
            case .failed(let name, let message):
                issues.append(SourceIssue(
                    kind: .network,
                    message: "\(name): \(message)",
                    recovery: "Check your network or package index and retry."))
            }
        }
        updates.sort { $0.name.localizedStandardCompare($1.name) == .orderedAscending }
        return SourceScanReport(updates: updates, issues: issues)
    }

    private func lookupPyPI(name: String, currentVersion: String) async -> PipxLookup {
        guard let encoded = name.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed),
              let url = URL(string: "https://pypi.org/pypi/\(encoded)/json") else {
            return .failed(name: name, message: "The package URL is invalid.")
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 15
        do {
            let response = try await httpClient.data(for: request)
            guard response.statusCode == 200 else {
                return .failed(name: name, message: "PyPI returned HTTP \(response.statusCode).")
            }
            let latest = try JSONDecoder().decode(PyPIResponse.self, from: response.data).info.version
            if !VersionComparator.isUpgrade(from: currentVersion, to: latest) { return .current(name: name) }
            return .outdated(PackageInfo(
                id: name,
                name: name,
                currentVersion: currentVersion,
                availableVersion: latest))
        } catch is CancellationError {
            return .failed(name: name, message: "Lookup was cancelled.")
        } catch {
            return .failed(name: name, message: error.decodingDescription)
        }
    }

    func clearCache(
        context: ToolContext,
        onOutput: @escaping @Sendable (ProcessOutputEvent) async -> Void
    ) async throws -> Int64 {
        let home = FileManager.default.homeDirectoryForCurrentUser
        let pipxCacheURLs = [
            home.appendingPathComponent("Library/Caches/pipx"),
            home.appendingPathComponent(".cache/pipx")
        ]
        let sizeBefore = SourceSupport.directorySize(at: pipxCacheURLs)

        let result = try await runner.run(
            context.executablePath,
            ["cache", "purge"],
            timeout: 180,
            environment: SourceSupport.environment(pathEntries: context.pathEntries),
            onOutput: onOutput)
        if !result.succeeded {
            let fallback = try await runner.run(
                context.executablePath,
                ["clear-cache"],
                timeout: 180,
                environment: SourceSupport.environment(pathEntries: context.pathEntries),
                onOutput: onOutput)
            guard fallback.succeeded else { throw SourceSupport.commandFailure("pipx cache purge", result: result) }
        }

        let sizeAfter = SourceSupport.directorySize(at: pipxCacheURLs)
        return max(0, sizeBefore - sizeAfter)
    }
}

enum PipxLookup: Sendable {
    case current(name: String)
    case outdated(PackageInfo)
    case failed(name: String, message: String)
}

struct PipxOutdatedEnvelope: Decodable {
    let status: String
    let exitCode: Int
    let data: OutdatedData
    let errors: [EnvelopeError]

    struct OutdatedData: Decodable {
        let packagesChecked: Int
        let packages: [Package]
        let skipped: [Skipped]
    }

    struct Package: Decodable {
        let environment: String
        let package: String
        let version: String
        let latestVersion: String
        let injected: Bool
        let pinned: Bool
    }

    struct Skipped: Decodable {
        let environment: String
        let package: String
        let reason: String
    }

    struct EnvelopeError: Decodable {
        let code: String
        let message: String
        let environment: String?
        let package: String?
    }
}

struct PyPIResponse: Decodable {
    let info: Info

    struct Info: Decodable {
        let version: String
    }
}

/// Recheck after a tool/version change or five minutes, without repeating help on every scan.
private actor PipxCapabilities {
    private var cache: [String: (checked: Date, supported: Bool)] = [:]

    func supportsNativeOutdated(context: ToolContext, runner: any ProcessRunning) async throws -> Bool {
        let attributes = try? FileManager.default.attributesOfItem(atPath: context.executablePath)
        let fingerprint = "\(attributes?[.modificationDate] ?? "")|\(attributes?[.size] ?? "")"
        let key = context.installationKey + "|" + fingerprint
        if let entry = cache[key], Date.now.timeIntervalSince(entry.checked) < 300 { return entry.supported }
        let result = try await runner.run(context.executablePath, ["list", "--help"], timeout: 15,
            environment: SourceSupport.environment(pathEntries: context.pathEntries))
        let supported = result.succeeded && result.stdout.contains("--outdated") && result.stdout.contains("--output")
        if result.succeeded {
            cache = cache.filter { Date.now.timeIntervalSince($0.value.checked) < 300 }
            cache[key] = (.now, supported)
        }
        return supported
    }
}

File: windows/src/PackMan/Services/PipxSource.cs
using System.Net;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PackMan.Services;

public sealed class PipxSource(IToolResolver resolver, IProcessRunner runner, HttpClient httpClient)
    : PackageSourceBase(resolver, runner)
{
    public override SourceDescriptor Descriptor { get; } = new(
        SourceId.Pipx, "pipx", ToolId.Pipx, "pipx",
        [Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Python", "Scripts", "pipx.exe")],
        "https://pipx.pypa.io/stable/installation/");

    private bool? _nativeOutdatedSupported;

    public override async Task<IReadOnlyDictionary<string, UpdateVerification>> VerifyAsync(
        IReadOnlyList<UpdateRequest> requests, ToolContext context, CancellationToken cancellationToken = default)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--short"), context.Environment, TimeSpan.FromMinutes(1)), cancellationToken: cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pipx installed inventory", result);
        var installed = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var line in result.StdOut.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length < 2 || !PackageIdValidator.IsValid(parts[0]) || !PackageIdValidator.IsValidVersion(parts[1]))
                throw new SourceException(SourceIssueKind.Verification, "pipx returned an unrecognized installed-package record.");
            installed[parts[0]] = parts[1];
        }
        return requests.ToDictionary(r => r.Identity, r => VerifyInstalled(installed.GetValueOrDefault(r.PackageId), r),
            StringComparer.OrdinalIgnoreCase);
    }

    public override bool SupportsCacheClear => true;

    public override async Task<string> ClearCacheAsync(ToolContext context,
        IProgress<ProcessOutputEvent>? output = null, CancellationToken cancellationToken = default)
    {
        // pipx delegates downloads to the pip inside each managed environment. Asking each
        // environment to purge its own configured pip cache also handles non-default locations.
        var list = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--short"), context.Environment, TimeSpan.FromMinutes(1)),
            cancellationToken: cancellationToken);
        if (!list.Success) throw SourceSupport.CommandFailure("pipx list", list);
        var packages = list.StdOut.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(line => line.Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault())
            .Where(PackageIdValidator.IsValid)
            .Select(package => package!)
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();
        if (packages.Count == 0) return "pipx has no managed environments with package caches.";

        var failures = new List<string>();
        foreach (var package in packages)
        {
            var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
                Arguments(context, "runpip", package, "cache", "purge"), context.Environment,
                TimeSpan.FromMinutes(2)), output, cancellationToken);
            if (!result.Success) failures.Add($"{package}: {SourceSupport.ErrorText(result)}");
        }
        if (failures.Count > 0)
            throw new SourceException(SourceIssueKind.Command,
                $"Could not clear {failures.Count} pipx cache{(failures.Count == 1 ? "" : "s")}: " +
                string.Join("; ", failures));
        return $"Cleared pip caches for {packages.Count} pipx environment{(packages.Count == 1 ? "" : "s")}.";
    }

    public override async Task<SourceScanReport> ScanAsync(ToolContext context,
        IProgress<SourcePhase>? progress = null, CancellationToken cancellationToken = default)
    {
        progress?.Report(SourcePhase.Scanning);
        var supported = _nativeOutdatedSupported;
        if (supported is null)
        {
            var help = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
                Arguments(context, "list", "--help"), context.Environment, TimeSpan.FromSeconds(20)),
                cancellationToken: cancellationToken);
            supported = help.Success && help.StdOut.Contains("--outdated") && help.StdOut.Contains("--output");
            _nativeOutdatedSupported = supported;
        }
        return supported.Value
            ? await ScanNativeAsync(context, cancellationToken)
            : await ScanLegacyAsync(context, cancellationToken);
    }

    public override async Task<UpdateResult> UpdateAsync(UpdateRequest request, ToolContext context,
        IProgress<ProcessOutputEvent>? output = null, CancellationToken cancellationToken = default)
    {
        Validate(request);
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "upgrade", request.PackageId), context.Environment, TimeSpan.FromMinutes(15), request.Elevated),
            output, cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pipx upgrade", result);
        return new();
    }

    private async Task<SourceScanReport> ScanNativeAsync(ToolContext context, CancellationToken cancellationToken)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--outdated", "--output", "json"), context.Environment, TimeSpan.FromMinutes(3)),
            cancellationToken: cancellationToken);
        if (string.IsNullOrWhiteSpace(result.StdOut))
        {
            if (result.Success) return SourceScanReport.Empty;
            throw SourceSupport.CommandFailure("pipx list --outdated", result);
        }
        PipxEnvelope envelope;
        try { envelope = JsonSerializer.Deserialize<PipxEnvelope>(result.StdOut) ?? new(); }
        catch (JsonException ex) { throw new SourceException(SourceIssueKind.Parsing, $"pipx JSON could not be parsed: {ex.Message}"); }
        var updates = (envelope.Data?.Packages ?? [])
            .Where(p => !p.Injected && !p.Pinned && PackageIdValidator.IsValid(p.Package))
            .Select(p => new PackageInfo(p.Package!, p.Package!, p.Version ?? string.Empty, p.LatestVersion ?? string.Empty)).ToList();
        var issues = (envelope.Errors ?? []).Select(error => new SourceIssue(SourceIssueKind.Network,
            $"{error.Environment ?? error.Package ?? "pipx"}: {error.Message ?? "outdated check failed"}",
            "Check the package index configuration and retry.")).ToList();
        if (!result.Success && issues.Count == 0)
            issues.Add(new(SourceIssueKind.Command, "pipx reported an unsuccessful outdated check.", SourceSupport.ErrorText(result)));
        return new(updates, issues);
    }

    private async Task<SourceScanReport> ScanLegacyAsync(ToolContext context, CancellationToken cancellationToken)
    {
        var result = await Runner.RunAsync(new ProcessInvocation(context.ExecutablePath,
            Arguments(context, "list", "--short"), context.Environment, TimeSpan.FromMinutes(1)),
            cancellationToken: cancellationToken);
        if (!result.Success) throw SourceSupport.CommandFailure("pipx list", result);
        var installed = new List<(string Name, string Version)>();
        var issues = new List<SourceIssue>();
        foreach (var line in result.StdOut.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length < 2 || !PackageIdValidator.IsValid(parts[0]))
                issues.Add(new(SourceIssueKind.Parsing, "Could not parse a pipx package record."));
            else installed.Add((parts[0], parts[1]));
        }
        using var gate = new SemaphoreSlim(4, 4);
        var lookups = await Task.WhenAll(installed.Select(async package =>
        {
            await gate.WaitAsync(cancellationToken);
            try { return await LookupPyPiAsync(package.Name, package.Version, cancellationToken); }
            finally { gate.Release(); }
        }));
        var updates = new List<PackageInfo>();
        foreach (var lookup in lookups)
        {
            if (lookup.Update is not null) updates.Add(lookup.Update);
            if (lookup.Issue is not null) issues.Add(lookup.Issue);
        }
        return new(updates.OrderBy(x => x.Name, StringComparer.CurrentCultureIgnoreCase).ToList(), issues);
    }

    private async Task<(PackageInfo? Update, SourceIssue? Issue)> LookupPyPiAsync(
        string name, string currentVersion, CancellationToken cancellationToken)
    {
        try
        {
            using var response = await httpClient.GetAsync($"https://pypi.org/pypi/{Uri.EscapeDataString(name)}/json", cancellationToken);
            if (response.StatusCode != HttpStatusCode.OK)
                return (null, new(SourceIssueKind.Network, $"{name}: PyPI returned HTTP {(int)response.StatusCode}."));
            await using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
            var payload = await JsonSerializer.DeserializeAsync<PyPiResponse>(stream, cancellationToken: cancellationToken);
            var latest = payload?.Info?.Version;
            if (string.IsNullOrWhiteSpace(latest)) return (null, new(SourceIssueKind.Parsing, $"{name}: PyPI returned no version."));
            return latest == currentVersion ? (null, null) : (new(name, name, currentVersion, latest), null);
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            return (null, new(SourceIssueKind.Network, $"{name}: {ex.Message}", "Check the network and retry."));
        }
    }

    private sealed class PipxEnvelope
    {
        [JsonPropertyName("data")] public PipxData? Data { get; set; }
        [JsonPropertyName("errors")] public List<PipxError>? Errors { get; set; }
    }
    private sealed class PipxData { [JsonPropertyName("packages")] public List<PipxPackage>? Packages { get; set; } }
    private sealed class PipxPackage
    {
        [JsonPropertyName("package")] public string? Package { get; set; }
        [JsonPropertyName("version")] public string? Version { get; set; }
        [JsonPropertyName("latest_version")] public string? LatestVersion { get; set; }
        [JsonPropertyName("injected")] public bool Injected { get; set; }
        [JsonPropertyName("pinned")] public bool Pinned { get; set; }
    }
    private sealed class PipxError
    {
        [JsonPropertyName("message")] public string? Message { get; set; }
        [JsonPropertyName("environment")] public string? Environment { get; set; }
        [JsonPropertyName("package")] public string? Package { get; set; }
    }
    private sealed class PyPiResponse { [JsonPropertyName("info")] public PyPiInfo? Info { get; set; } }
    private sealed class PyPiInfo { [JsonPropertyName("version")] public string? Version { get; set; } }
}
